src/hashtable.c

```diff
@@ -52,12 +59,39 @@ Hashtable* hashtable_new(size_t size, HTCmpFunc cmp_func, HTHashFunc hash_func,
     ht->hash_func = hash_func;
     ht->copy_func = copy_func;
     ht->free_func = free_func;
-    ht->deferred_unsets = NULL;
 
+    SDL_AtomicSet(&ht->cur_operation, HT_OP_NONE);
     assert(ht->hash_func != NULL);
 
     return ht;
 }
+static bool hashtable_try_enter_state(Hashtable *ht, int state, bool mutex) {
+    SDL_atomic_t *ptr = &ht->cur_operation;
+
+    if(!mutex) {
+        SDL_AtomicCAS(ptr, state, HT_OP_NONE);
+    }
+
+    return SDL_AtomicCAS(ptr, HT_OP_NONE, state);
+}
+
+static void hashtable_enter_state(Hashtable *ht, int state, bool mutex) {
+    while(!hashtable_try_enter_state(ht, state, mutex));
+}
+
+static void hashtable_idle_state(Hashtable *ht) {
+    SDL_AtomicSet(&ht->cur_operation, HT_OP_NONE);
+}
+
+void hashtable_lock(Hashtable *ht) {
```

Comment:
It stands for "mutual exclusion". If `mutex` is true, the call to `hashtable_enter_state` will block until the state is `HT_OP_NONE` (effectively implementing a [critical section](https://en.wikipedia.org/wiki/Critical_section)). If it's false, it will also do that, but only if the current state is different from the state we want to enter (and is not `HT_OP_NONE`, of course).

`mutex` is true for operations that modify the hashtable, and false for those that only read from it. This allows multiple threads to read from the hashtable at the same time, but forbids concurrent writes, reads during writes, and writes during reads. Since reading from a hashtable ought to be more common than writing to it, I think it's good to get rid of unnecessary blocking there.

I previously used [SDL's mutexes](https://wiki.libsdl.org/SDL_CreateMutex) there for synchronization, but they aren't flexible enough to enable the exact behaviour I wanted (concurrent reads), so I rolled my own version using the atomic primitives.